### PR TITLE
API: ensure read_json closes file handle

### PR DESCRIPTION
**Pull request by topper-123**
Use context manager instead of closing manually with `self.close` to ensure file always closes.

Not sure if this should have a whatsnew entry and where.

### CODE REVIEW

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me (easier to review with the "hide whitespace" option)

**Comment by topper-123** [comment]
> easier to review with the "hide whitespace" option

Nice one, I didn't know about that one. It does indeed makes changes that are mostly changing the indention level easier to review.

**Comment by MarcoGorelli** [comment]
Yeah I don't think this needs a whatsnew entry - leaving open for a bit in case others have comments / disagree

**Comment by mroeschke** [comment]
Could note in the `Other API changes` that `read_json` will always close the file if there was an error in reading the file?

**Comment by topper-123** [comment]
Thanks for the comments.

I've added a whatsnew entry + updated the PR to also improve the handling of closing files in `JsonReader.__next__` also.

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -349,6 +349,7 @@ Other API changes
 - :func:`read_stata` with parameter ``index_col`` set to ``None`` (the default) will now set the index on the returned :class:`DataFrame` to a :class:`RangeIndex` instead of a :class:`Int64Index` (:issue:`49745`)
 - Changed behavior of :class:`Index` constructor with an object-dtype ``numpy.ndarray`` containing all-``bool`` values or all-complex values, this will now retain object dtype, consistent with the :class:`Series` behavior (:issue:`49594`)
 - Changed behavior of :meth:`DataFrame.shift` with ``axis=1``, an integer ``fill_value``, and homogeneous datetime-like dtype, this now fills new columns with integer dtypes instead of casting to datetimelike (:issue:`49842`)
+- Improve handling of file closures in :func:`read_json` (:issue:`49921`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:352]:
```suggestion
- Files are now closed when encountering an exception in :func:`read_json` (:issue:`49921`)
```

**Review by mroeschke — COMMENTED** [review]
Would it be easy to test that the files handles are closed  when an exception is encountered?

**Comment by topper-123** [comment]
> Would it be easy to test that the files handles are closed when an exception is encountered?

not sure, maybe serving a malformed file (though haven't checked if that is tested already or maybe caught in the init)?

**Comment by topper-123** [comment]
error looks unrelated ("ERROR: Could not find a version that satisfies the requirement setuptools==59.2.0")

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @topper-123. If you have any inspirations on how to test that files are closed on exceptions happy to have a follow up

**Pull request merged**